crates/hir_ty/src/diagnostics/decl_check/str_helpers.rs

```diff
@@ -0,0 +1,195 @@
+        DetectedCase::LowerCamelCase => {
+            let mut first_capitalized = false;
+            let output = ident
+                .chars()
+                .map(|chr| {
+                    if !first_capitalized && chr.is_ascii_lowercase() {
+                        first_capitalized = true;
+                        chr.to_ascii_uppercase()
+                    } else {
+                        chr
+                    }
+                })
+                .collect();
+            return Some(output);
+        }
+        _ => {}
+    }
+
+    let mut output = String::with_capacity(ident.len());
+
+    let mut capital_added = false;
+    for chr in ident.chars() {
+        if chr.is_alphabetic() {
+            if !capital_added {
+                output.push(chr.to_ascii_uppercase());
+                capital_added = true;
+            } else {
+                output.push(chr.to_ascii_lowercase());
+            }
+        } else if chr == '_' {
+            // Skip this character and make the next one capital.
+            capital_added = false;
+        } else {
+            // Put the characted as-is.
+            output.push(chr);
+        }
+    }
+
+    if output == ident {
+        // While we didn't detect the correct case at the beginning, there
+        // may be special cases: e.g. `A` is both valid CamelCase and UPPER_SNAKE_CASE.
+        None
+    } else {
+        Some(output)
+    }
+}
+
+/// Converts an identifier to a lower_snake_case form.
+/// Returns `None` if the string is already in lower_snake_case.
+pub fn to_lower_snake_case(ident: &str) -> Option<String> {
+    // First, assume that it's UPPER_SNAKE_CASE.
+    match detect_case(ident) {
+        DetectedCase::LowerSnakeCase => return None,
+        DetectedCase::UpperSnakeCase => {
+            return Some(ident.chars().map(|chr| chr.to_ascii_lowercase()).collect())
+        }
+        _ => {}
+    }
+
+    // Otherwise, assume that it's CamelCase.
+    let lower_snake_case = stdx::to_lower_snake_case(ident);
+
+    if lower_snake_case == ident {
+        // While we didn't detect the correct case at the beginning, there
+        // may be special cases: e.g. `a` is both valid camelCase and snake_case.
+        None
+    } else {
+        Some(lower_snake_case)
+    }
+}
+
+/// Converts an identifier to an UPPER_SNAKE_CASE form.
+/// Returns `None` if the string is already is UPPER_SNAKE_CASE.
+pub fn to_upper_snake_case(ident: &str) -> Option<String> {
+    match detect_case(ident) {
+        DetectedCase::UpperSnakeCase => return None,
+        DetectedCase::LowerSnakeCase => {
+            return Some(ident.chars().map(|chr| chr.to_ascii_uppercase()).collect())
+        }
+        _ => {}
+    }
+
+    // Normalize the string from whatever form it's in currently, and then just make it uppercase.
+    let upper_snake_case =
+        stdx::to_lower_snake_case(ident).chars().map(|c| c.to_ascii_uppercase()).collect();
```

Comment:
maybe just add `to_upper_snake_case` to stdx?

Reply:
I was a bit hesitant to add a function which is (currently) is only used in one place into a dependency crate, but surely it will fit there better :)